### PR TITLE
enable multi-threading in onnx-mlir

### PR DESCRIPTION
**Pull request by lixi-zhou**
Enable the multi-threading in onnx-mlir by utilizing OpenMP.  Tested on x86 with performance gain. 

Introduced a new operator in krnl dialect: krnl.parallel which takes a loop as input to specify the loop will be parallelized. 

To parallelization is **enabled** with the argument `--parallel` in `onnx-mlir` and  `--convert-onnx-to-krnl=enable-parallel` in `onnx-mlir-opt`.

To parallel an operator while lowering, you can follow the attached example below: 

```diff
ValueRange loopDef = create.krnl.defineLoops(outputRank);
SmallVector<IndexExpr, 4> lbs(outputRank, LiteralIndexExpr(0));
SmallVector<IndexExpr, 4> ubs;
+ create.krnl.parallel(loopDef[0]) // parallelize the outermost loop
create.krnl.iterateIE(loopDef, loopDef, lbs, ubs,
          [&](KrnlBuilder &createKrnl, ValueRange loopInd) {
      ...
}
```

To run your model with parallelization, please build your MLIR along with OpenMP with following code and set the dependency with `export LD_PRELOAD=../llvm-project/build/lib/libomp.so`.

```diff
+ DLLVM_ENABLE_PROJECTS="mlir;openmp;clang" 
- DLLVM_ENABLE_PROJECTS="mlir" 
```

The parallelization is done by transforming the affine.forOp to affine.parallelOp this happens in the Pass:  `convert-krnl-to-affine`. krnl.parallel is interpreted after krnl.defineloop, krnl.iterate, krnl.block, krnl.permute and krnl.unroll. The further lowering will be handled by `convert-scf-to-openmp` and `convert-openmp-to-llvm` automatically.

Current parallelized operators (outermost loop):
- [x] Elementwise Operator
- [x] GEMM
- [x] MatMul
- [x] Transpose
- [x] Softmax
- [x] Reduction
- [x] Concat

Signed-off-by: @lixi-zhou 
Co-authored-by:  @AlexandreEichenberger

### CODE REVIEW

**Comment by AlexandreEichenberger** [comment]
@lixi-zhou 

The errors in the build occurs in Release mode; in your test env, change / add this I believe. ` -DCMAKE_BUILD_TYPE=Release `. Default (without it) is debug

File: src/Conversion/ONNXToKrnl/Math/Elementwise.cpp [hunk 1]
```diff
@@ -2233,6 +2252,11 @@ struct ONNXElementwiseVariadicOpLowering
       SmallVector<IndexExpr, 4> ubs;
       create.krnlIE.getShapeAsDims(alloc, ubs);
 
+      // FIXME fix in the future
```
Comment (AlexandreEichenberger) [inline src/Conversion/ONNXToKrnl/Math/Elementwise.cpp:2255]: Fix are fine, please states what needs to be fixed.

**Review by AlexandreEichenberger — COMMENTED** [review]
First batch of comments, will look next more closely where the parallel is introduced.

**Comment by lixi-zhou** [comment]
@jenkins-droid  test this please

File: src/Conversion/ONNXToKrnl/Math/Reduction.cpp [hunk 2]
```diff
@@ -722,6 +734,10 @@ struct ONNXReductionOpLowering : public OpConversionPattern<ONNXReductionOp> {
     SmallVector<IndexExpr, 4> ubs2;
     create.krnlIE.getShapeAsSymbols(input, ubs2);
     Value trueVal = create.math.constant(rewriter.getIntegerType(1), 1);
+    if (enableParallel) {
```
Comment (AlexandreEichenberger) [inline src/Conversion/ONNXToKrnl/Math/Reduction.cpp:737]: I believe that in the generic case, the outermost loop may be a reduction loop, in which case enabling parallelism would not be safe.

Either this is tested at compile time (namely the reduction pattern rules out outermost loop to be a reduction loop), or it is simply disabled in all cases.

The first and last loop nest are fine, just this one.

File: src/Conversion/ONNXToKrnl/Math/Softmax.cpp [hunk 3]
```diff
@@ -166,6 +168,10 @@ void emitInstForSoftmax<ONNXSoftmaxV11Op>(ConversionPatternRewriter &rewriter,
     SmallVector<IndexExpr, 4> outerUbs;
     for (int i = 0; i < axis; ++i)
       outerUbs.emplace_back(create.krnlIE.getShapeAsDim(input, i));
+    if (enableParallel) {
```
Comment (AlexandreEichenberger) [inline src/Conversion/ONNXToKrnl/Math/Softmax.cpp:171]: For softmax, don't we have to check that the outer loop is not one on which the softmax operation is performed?

```
Attributes
axis - INT (default is '-1'): Describes the dimension Softmax will be performed on. Negative value means counting dimensions from the back. Accepted range is [-r, r-1] where r = rank(input).
```

Please double check.

File: src/Dialect/Krnl/Krnl.td [hunk 4]
```diff
@@ -500,6 +500,23 @@ def KrnlUnrollOp : Op<Krnl_Dialect, "unroll"> {
   }];
 }
 
+def KrnlParallelOp : Op<Krnl_Dialect, "parallel"> {
+  let summary = "Krnl parallel operation";
+  let description = [{
+    Parallelize the specified loops.
+    ```
+    krnl.parallel %i
```
Comment (AlexandreEichenberger) [inline src/Dialect/Krnl/Krnl.td:508]: I believe you have constrains... e.g. it has to be the last operation before the krill.iterate, isn't it? If so, I would mention this here. For example, we don't want to parallelize and the block that loop. Or permute?

File: src/Conversion/KrnlToAffine/ConvertKrnlToAffine.cpp [hunk 5]
```diff
@@ -334,21 +340,26 @@ static void markLoopBodyAsMovable(
...
+    llvm::SmallDenseMap<Value, Operation *, 4> &loopRefToOp) {
   auto zippedOperandsResults =
       llvm::zip(getIVOp->getOperands(), getIVOp->getResults());
   for (const auto &operandAndResult : zippedOperandsResults) {
     auto operand = std::get<0>(operandAndResult);
     auto result = std::get<1>(operandAndResult);
-    result.replaceAllUsesWith(loopRefToOp[operand].getInductionVar());
+    if (auto forOp =  dyn_cast_or_null<AffineForOp>(loopRefToOp[operand])) {
+      result.replaceAllUsesWith(forOp.getInductionVar());
+    } else if (auto parallelOp =
```
[the remark this answers is not in the document]
Comment (lixi-zhou) [reply src/Conversion/KrnlToAffine/ConvertKrnlToAffine.cpp]: Have addressed in the recent commit.

File: src/Conversion/KrnlToAffine/ConvertKrnlToAffine.cpp [hunk 6]
```diff
@@ -523,7 +540,7 @@ static LogicalResult interpretOperation(Operation *op, OpBuilder &builder,
     // for loops in loopRefToLoop.
     loopRefToOp.erase(loopRefToOp.find_as(blockOp.getLoop()));
     loopRefToOp[blockOp.getResult(0)] = tiledLoops[0];
-    loopRefToOp[blockOp.getResult(1)] = tiledLoops[1];
+    loopRefToOp[blockOp.getResult(1)] = tiledLoops[1];
```
[the remark this answers is not in the document]
Comment (lixi-zhou) [reply src/Conversion/KrnlToAffine/ConvertKrnlToAffine.cpp]: run clangformat on the files, and it gets passed in the recent commit

File: src/Conversion/KrnlToAffine/ConvertKrnlToAffine.cpp [hunk 7]
```diff
@@ -581,8 +600,62 @@ static LogicalResult interpretOperation(Operation *op, OpBuilder &builder,
...
     opsToErase.insert(op);
     return success();
-  }
+  } else if (auto parallelOp = dyn_cast_or_null<KrnlParallelOp>(op)) {
+    // Parallelism the given loop by transform the tagged affine.for op to 
+    // affine.parallel
+    LLVM_DEBUG(llvm::dbgs() << DEBUG_TYPE 
+              << " interpret parallel op " << parallelOp << "\n");
+
+    Value loopRef = parallelOp.getLoop();
```
[the remark this answers is not in the document]
Comment (lixi-zhou) [reply src/Conversion/KrnlToAffine/ConvertKrnlToAffine.cpp:615]: Comment is added in recent commit.

File: src/Conversion/ONNXToKrnl/Math/Elementwise.cpp (hunk 1 above)
Comment (lixi-zhou) [reply src/Conversion/ONNXToKrnl/Math/Elementwise.cpp]: Removed it, there is nothing needed to be fixed here

**Comment by lixi-zhou** [comment]
Closed the current PR and created a new one, #2435, to continue the code review discussion because of a weird failure on Jenkins CI.

**Comment by AlexandreEichenberger** [comment]
continued as #2435

**Pull request closed without merging**